core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -252,6 +257,23 @@ private void ensureNoPre019ShardState(NodeEnvironment nodeEnv) throws Exception
         }
     }
 
+    private void upgradePluginsMetaData() throws Exception {
+        MetaData metaData = loadMetaState();
+        boolean changed = false;
+        for (Function<MetaData, MetaData> metaDataUpgrader : pluginsService.getMetaDataUpgrader()) {
+            MetaData upgradedMetaData = metaDataUpgrader.apply(metaData);
```

Comment:
I wonder if we should either check for `Function.identity()` or allow the function to return `null` for "no change needed".